File: csharp/src/Toolnexus/InProcess.cs
using System.Text;
using System.Text.Json;

namespace Toolnexus;

/// <summary>
/// In-process models (SPEC §8 Gap 2, semantic form).
///
/// <para>A model running in this process is not a network endpoint, so it should not need a base
/// URL, an API key, a style, or an <see cref="HttpMessageHandler"/>. You supply
/// <c>Generate</c>; everything HTTP-shaped is built here.</para>
///
/// <para>This is a second constructor, not a second seam: it is built on the shipped injectable
/// transport, so the tool-calling loop, MCP servers, skills, sub-agents, hooks, metrics and the
/// completion gate behave identically.</para>
/// </summary>
public static class InProcess
{
    /// <summary>
    /// One tool call an in-process model asks for. Flat on purpose: the nested <c>function:{}</c>
    /// wrapper is a wire detail. <c>Arguments</c> may be any value (encoded for you) or an
    /// already-encoded string, passed through unchanged.
    /// </summary>
    public sealed class ToolCall
    {
        public string? Id { get; set; }
        public string Name { get; set; } = "";
        public object? Arguments { get; set; }
    }

    /// <summary>Optional token reporting. Absent ⇒ the run reports zero rather than failing.</summary>
    public sealed class Usage
    {
        public int PromptTokens { get; set; }
        public int CompletionTokens { get; set; }
        public int TotalTokens { get; set; }
    }

    /// <summary>What an in-process model is handed: the assembled request for THIS call.</summary>
    public sealed class Request
    {
        public IReadOnlyList<object?> Messages { get; init; } = Array.Empty<object?>();
        public IReadOnlyList<object?> Tools { get; init; } = Array.Empty<object?>();
        public string Model { get; init; } = "";
        /// <summary>Every key the client assembled, for a model that wants them.</summary>
        public IReadOnlyDictionary<string, object?> Body { get; init; } =
            new Dictionary<string, object?>();
    }

    /// <summary>Exactly one assistant message. Set Content to finish, or ToolCalls to ask for tools.</summary>
    public sealed class Response
    {
        public string? Content { get; set; }
        public List<ToolCall>? ToolCalls { get; set; }
        public Usage? Usage { get; set; }

        public static Response FromContent(string text) => new() { Content = text };
        public static Response FromToolCalls(params ToolCall[] calls) => new() { ToolCalls = calls.ToList() };
    }

    /// <summary>Every client option EXCEPT the three that describe a wire, plus <c>Generate</c>.</summary>
    public sealed class Options
    {
        public string Model { get; set; } = "";
        public Func<Request, Response>? Generate { get; set; }
        public string? SystemPrompt { get; set; }
        public int? MaxTurns { get; set; }
        public LlmClient.Hooks? Hooks { get; set; }
        public long? TimeoutMs { get; set; }
        public IConversationStore? Store { get; set; }
        public Action<MetricEvent>? OnMetric { get; set; }
        public Func<Toolnexus.Request, Task<Answer>>? WaitFor { get; set; }
        public IDictionary<string, object?>? RequestParams { get; set; }
        /// <summary>Defaults to 0 for an in-process client. Set only if your model is genuinely flaky.</summary>
        public int? Retries { get; set; }
    }

    /// <summary>
    /// A sentinel. Never dialled — the handler answers every request before the network is reached —
    /// but a URL string is built internally, so it must be syntactically valid. <c>.invalid</c> is
    /// reserved by RFC 2606 precisely so a name can never resolve.
    /// </summary>
    private const string BaseUrl = "http://in-process.invalid/v1";

    /// <summary>Build a client backed by a model running IN THIS PROCESS.</summary>
    public static LlmClient CreateClient(Options opts)
    {
        if (opts?.Generate is null)
            throw new ArgumentException("toolnexus: InProcess.CreateClient requires a `Generate` function");

        var o = new LlmClient.Options
        {
            BaseUrl = BaseUrl,
            Style = "openai",
            Model = opts.Model,
            HttpHandler = new GenerateBackedHandler(opts.Generate),
            SystemPrompt = opts.SystemPrompt,
            Hooks = opts.Hooks,
            Store = opts.Store,
            OnMetric = opts.OnMetric,
            WaitFor = opts.WaitFor,
            RequestParams = opts.RequestParams is null ? null : new Dictionary<string, object?>(opts.RequestParams),
            // Zero retries by default: there is no wire, so there is no transient failure to ride
            // out, and retrying only buys backoff before the caller sees their own bug.
            Retries = opts.Retries ?? 0,
        };
        // Only when set: a 0 here is not "no timeout", it is "already expired".
        if (opts.MaxTurns is > 0) o.MaxTurns = opts.MaxTurns.Value;
        if (opts.TimeoutMs is > 0) o.TimeoutMs = opts.TimeoutMs.Value;
        return LlmClient.Create(o);
    }

    private sealed class GenerateBackedHandler : HttpMessageHandler
    {
        private readonly Func<Request, Response> _generate;
        public GenerateBackedHandler(Func<Request, Response> generate) => _generate = generate;

        protected override async Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request, CancellationToken cancellationToken)
        {
            var raw = request.Content is null
                ? "{}" : await request.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            using var doc = JsonDocument.Parse(string.IsNullOrWhiteSpace(raw) ? "{}" : raw);
            var root = doc.RootElement;

            // A Generate returns a whole answer, so a stream would be one chunk pretending to be
            // many — indistinguishable from a real stream by content or delta count. Refuse.
            if (root.TryGetProperty("stream", out var s) && s.ValueKind == JsonValueKind.True)
                throw new NotSupportedException(
                    "toolnexus: InProcess.CreateClient does not support streaming — `Generate` returns "
                    + "a complete answer. Use RunAsync, or supply an HttpHandler that streams.");

            var answer = _generate(new Request
            {
                Messages = ToList(root, "messages"),
                Tools = ToList(root, "tools"),
                Model = root.TryGetProperty("model", out var m) ? m.GetString() ?? "" : "",
                Body = JsonSerializer.Deserialize<Dictionary<string, object?>>(raw)
                       ?? new Dictionary<string, object?>(),
            }) ?? Response.FromContent("");

            var sb = new StringBuilder();
            sb.Append("{\"choices\":[{\"index\":0,\"message\":{\"role\":\"assistant\"");
            if (answer.ToolCalls is { Count: > 0 })
            {
                sb.Append(",\"tool_calls\":[");
                for (var i = 0; i < answer.ToolCalls.Count; i++)
                {
                    var c = answer.ToolCalls[i];
                    if (i > 0) sb.Append(',');
                    sb.Append("{\"id\":").Append(JsonSerializer.Serialize(c.Id ?? $"call_{i}"))
                      .Append(",\"type\":\"function\",\"function\":{\"name\":")
                      .Append(JsonSerializer.Serialize(c.Name))
                      .Append(",\"arguments\":").Append(JsonSerializer.Serialize(EncodeArgs(c.Arguments)))
                      .Append("}}");
                }
                sb.Append(']');
            }
            else
            {
                sb.Append(",\"content\":").Append(JsonSerializer.Serialize(answer.Content ?? ""));
            }
            var finish = answer.ToolCalls is { Count: > 0 } ? "tool_calls" : "stop";

            var prompt = answer.Usage?.PromptTokens ?? 0;
            var completion = answer.Usage?.CompletionTokens ?? 0;
            var total = answer.Usage?.TotalTokens is > 0 ? answer.Usage!.TotalTokens : prompt + completion;

            sb.Append("},\"finish_reason\":\"").Append(finish).Append("\"}],\"usage\":{")
              .Append("\"prompt_tokens\":").Append(prompt)
              .Append(",\"completion_tokens\":").Append(completion)
              .Append(",\"total_tokens\":").Append(total).Append("}}");

            return new HttpResponseMessage(System.Net.HttpStatusCode.OK)
            {
                Content = new StringContent(sb.ToString(), Encoding.UTF8, "application/json"),
            };
        }

        private static IReadOnlyList<object?> ToList(JsonElement root, string name) =>
            root.TryGetProperty(name, out var el) && el.ValueKind == JsonValueKind.Array
                ? JsonSerializer.Deserialize<List<object?>>(el.GetRawText()) ?? new List<object?>()
                : new List<object?>();

        /// <summary>Pass an already-encoded string through; encode anything else.</summary>
        private static string EncodeArgs(object? v) => v switch
        {
            null => "{}",
            string s => s,
            _ => JsonSerializer.Serialize(v),
        };
    }
}

File: java/src/main/java/io/github/muthuishere/toolnexus/InProcess.java
package io.github.muthuishere.toolnexus;

import java.io.IOException;
import java.io.InputStream;
import java.net.Authenticator;
import java.net.CookieHandler;
import java.net.ProxySelector;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpHeaders;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.function.Function;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.SSLSession;

/**
 * In-process models (SPEC §8 Gap 2, semantic form).
 *
 * <p>A model running in this JVM is not a network endpoint, so it should not need a base URL, an
 * API key, a style, or — the real tax in this port — a 94-line {@link HttpClient} subclass. You
 * supply {@code generate}; everything HTTP-shaped is built here.
 *
 * <p>This is a second constructor, not a second seam: it is built on the shipped injectable
 * {@code httpClient}, so the tool-calling loop, MCP servers, skills, sub-agents, hooks, metrics and
 * the completion gate behave identically.
 */
public final class InProcess {

    private InProcess() {}

    /**
     * One tool call an in-process model asks for. Flat on purpose: the nested {@code function:{}}
     * wrapper is a wire detail, not something a model author should type. {@code arguments} may be
     * any value (encoded for you) or an already-encoded String, passed through unchanged.
     */
    public static final class ToolCall {
        public String id;
        public String name;
        public Object arguments;

        public ToolCall(String name, Object arguments) { this.name = name; this.arguments = arguments; }
        public ToolCall id(String v) { this.id = v; return this; }
    }

    /** Optional token reporting. Absent ⇒ the run reports zero rather than failing. */
    public static final class Usage {
        public int promptTokens;
        public int completionTokens;
        public int totalTokens;

        public Usage(int promptTokens, int completionTokens) {
            this.promptTokens = promptTokens;
            this.completionTokens = completionTokens;
        }
    }

    /** What an in-process model is handed: the assembled request for THIS call. */
    public static final class Request {
        public final List<Object> messages;
        public final List<Object> tools;
        public final String model;
        /** Every key the client assembled, for a model that wants them. */
        public final Map<String, Object> body;

        Request(List<Object> messages, List<Object> tools, String model, Map<String, Object> body) {
            this.messages = messages;
            this.tools = tools;
            this.model = model;
            this.body = body;
        }
    }

    /** Exactly one assistant message. Set content to finish, or toolCalls to ask for tools. */
    public static final class Response {
        public String content;
        public List<ToolCall> toolCalls;
        public Usage usage;

        public static Response content(String text) {
            Response r = new Response();
            r.content = text;
            return r;
        }

        public static Response toolCalls(ToolCall... calls) {
            Response r = new Response();
            r.toolCalls = List.of(calls);
            return r;
        }

        public Response usage(int promptTokens, int completionTokens) {
            this.usage = new Usage(promptTokens, completionTokens);
            return this;
        }
    }

    /** Every client option EXCEPT the three that describe a wire, plus {@code generate}. */
    public static final class Options {
        public String model;
        public Function<Request, Response> generate;
        public String systemPrompt;
        public int maxTurns;
        public LlmClient.Hooks hooks;
        public long timeoutMs;
        public LlmClient.ConversationStore store;
        public java.util.function.Consumer<LlmClient.MetricEvent> onMetric;
        public Function<io.github.muthuishere.toolnexus.Request, Answer> waitFor;
        public Map<String, Object> requestParams;

        public Options model(String v) { this.model = v; return this; }
        public Options generate(Function<Request, Response> v) { this.generate = v; return this; }
        public Options systemPrompt(String v) { this.systemPrompt = v; return this; }
        public Options maxTurns(int v) { this.maxTurns = v; return this; }
        public Options hooks(LlmClient.Hooks v) { this.hooks = v; return this; }
        public Options timeoutMs(long v) { this.timeoutMs = v; return this; }
        public Options store(LlmClient.ConversationStore v) { this.store = v; return this; }
        public Options onMetric(java.util.function.Consumer<LlmClient.MetricEvent> v) { this.onMetric = v; return this; }
        public Options requestParams(Map<String, Object> v) { this.requestParams = v; return this; }
    }

    /**
     * A sentinel. Never dialled — the client below answers every request before the network is
     * reached — but a URL string is built internally, so it must be syntactically valid.
     * {@code .invalid} is reserved by RFC 2606 precisely so a name can never resolve.
     */
    private static final String BASE_URL = "http://in-process.invalid/v1";

    /** Build a client backed by a model running IN THIS PROCESS. */
    public static LlmClient createClient(Options opts) {
        if (opts == null || opts.generate == null) {
            throw new IllegalArgumentException("toolnexus: InProcess.createClient requires a `generate` function");
        }
        LlmClient.Options o = new LlmClient.Options();
        o.baseUrl = BASE_URL;
        o.style = "openai";
        o.model = opts.model;
        o.httpClient = new GenerateBackedHttpClient(opts.generate);
        o.systemPrompt = opts.systemPrompt;
        if (opts.maxTurns > 0) o.maxTurns = opts.maxTurns;
        o.hooks = opts.hooks;
        // Only when positive: 0 is not "no timeout" here, it is "already expired".
        if (opts.timeoutMs > 0) o.timeoutMs = opts.timeoutMs;
        o.store = opts.store;
        o.onMetric = opts.onMetric;
        o.waitFor = opts.waitFor;
        o.requestParams = opts.requestParams;
        return LlmClient.create(o);
    }

    // ---- the HTTP shim the host no longer has to write ---------------------------

    static final class GenerateBackedHttpClient extends HttpClient {
        private final Function<Request, Response> generate;

        GenerateBackedHttpClient(Function<Request, Response> generate) { this.generate = generate; }

        @Override @SuppressWarnings("unchecked")
        public <T> HttpResponse<T> send(HttpRequest req, HttpResponse.BodyHandler<T> handler) throws IOException {
            Map<String, Object> body = readBody(req);

            // A generate returns a whole answer, so a stream would be one chunk pretending to be
            // many — indistinguishable from a real stream by content or delta count. Refuse.
            if (Boolean.TRUE.equals(body.get("stream"))) {
                throw new IOException("toolnexus: InProcess.createClient does not support streaming — "
                        + "`generate` returns a complete answer. Use run(), or supply an httpClient that streams.");
            }

            List<Object> messages = asList(body.get("messages"));
            List<Object> tools = asList(body.get("tools"));
            Response answer = generate.apply(
                    new Request(messages, tools, String.valueOf(body.get("model")), body));
            if (answer == null) answer = Response.content("");

            Map<String, Object> message = new LinkedHashMap<>();
            message.put("role", "assistant");
            if (answer.toolCalls != null && !answer.toolCalls.isEmpty()) {
                List<Object> calls = new ArrayList<>();
                for (int i = 0; i < answer.toolCalls.size(); i++) {
                    ToolCall c = answer.toolCalls.get(i);
                    Map<String, Object> fn = new LinkedHashMap<>();
                    fn.put("name", c.name);
                    fn.put("arguments", encodeArgs(c.arguments));
                    Map<String, Object> call = new LinkedHashMap<>();
                    call.put("id", c.id != null ? c.id : "call_" + i);
                    call.put("type", "function");
                    call.put("function", fn);
                    calls.add(call);
                }
                message.put("tool_calls", calls);
            } else {
                message.put("content", answer.content != null ? answer.content : "");
            }

            int prompt = answer.usage != null ? answer.usage.promptTokens : 0;
            int completion = answer.usage != null ? answer.usage.completionTokens : 0;
            int total = answer.usage != null && answer.usage.totalTokens > 0
                    ? answer.usage.totalTokens : prompt + completion;

            Map<String, Object> choice = new LinkedHashMap<>();
            choice.put("index", 0);
            choice.put("message", message);
            choice.put("finish_reason", message.containsKey("tool_calls") ? "tool_calls" : "stop");

            Map<String, Object> usage = new LinkedHashMap<>();
            usage.put("prompt_tokens", prompt);
            usage.put("completion_tokens", completion);
            usage.put("total_tokens", total);

            Map<String, Object> envelope = new LinkedHashMap<>();
            envelope.put("choices", List.of(choice));
            envelope.put("usage", usage);

            return (HttpResponse<T>) new JsonResponse(req, Json.stringify(envelope));
        }

        @SuppressWarnings("unchecked")
        private static Map<String, Object> readBody(HttpRequest req) {
            // The client always sends a JSON body; re-reading it here is how the shim stays
            // behind the shipped seam instead of needing a new one.
            java.util.Optional<HttpRequest.BodyPublisher> bp = req.bodyPublisher();
            if (bp.isEmpty()) return Map.of();
            StringBuilder sb = new StringBuilder();
            bp.get().subscribe(new java.util.concurrent.Flow.Subscriber<>() {
                @Override public void onSubscribe(java.util.concurrent.Flow.Subscription s) { s.request(Long.MAX_VALUE); }
                @Override public void onNext(java.nio.ByteBuffer item) { sb.append(java.nio.charset.StandardCharsets.UTF_8.decode(item)); }
                @Override public void onError(Throwable t) {}
                @Override public void onComplete() {}
            });
            Map<String, Object> parsed = Json.toMap(sb.toString());
            return parsed != null ? parsed : Map.of();
        }

        @SuppressWarnings("unchecked")
        private static List<Object> asList(Object v) {
            return v instanceof List ? (List<Object>) v : List.of();
        }

        /** Pass an already-encoded string through; encode anything else. */
        private static String encodeArgs(Object v) {
            if (v == null) return "{}";
            if (v instanceof String s) return s;
            return Json.stringify(v);
        }

        // Only send() is ever called by the client; the rest exist because HttpClient is abstract.
        @Override public Optional<CookieHandler> cookieHandler() { return Optional.empty(); }
        @Override public Optional<Duration> connectTimeout() { return Optional.empty(); }
        @Override public Redirect followRedirects() { return Redirect.NEVER; }
        @Override public Optional<ProxySelector> proxy() { return Optional.empty(); }
        @Override public SSLContext sslContext() { throw new UnsupportedOperationException("in-process client"); }
        @Override public SSLParameters sslParameters() { throw new UnsupportedOperationException("in-process client"); }
        @Override public Optional<Authenticator> authenticator() { return Optional.empty(); }
        @Override public Version version() { return Version.HTTP_1_1; }
        @Override public Optional<Executor> executor() { return Optional.empty(); }

        @Override public <T> CompletableFuture<HttpResponse<T>> sendAsync(
                HttpRequest r, HttpResponse.BodyHandler<T> h) {
            try {
                return CompletableFuture.completedFuture(send(r, h));
            } catch (IOException e) {
                return CompletableFuture.failedFuture(e);
            }
        }

        @Override public <T> CompletableFuture<HttpResponse<T>> sendAsync(
                HttpRequest r, HttpResponse.BodyHandler<T> h, HttpResponse.PushPromiseHandler<T> p) {
            return sendAsync(r, h);
        }
    }

    record JsonResponse(HttpRequest req, String payload) implements HttpResponse<String> {
        @Override public int statusCode() { return 200; }
        @Override public String body() { return payload; }
        @Override public HttpRequest request() { return req; }
        @Override public HttpHeaders headers() {
            return HttpHeaders.of(Map.of("content-type", List.of("application/json")), (a, b) -> true);
        }
        @Override public Optional<HttpResponse<String>> previousResponse() { return Optional.empty(); }
        @Override public Optional<SSLSession> sslSession() { return Optional.empty(); }
        @Override public URI uri() { return req.uri(); }
        @Override public HttpClient.Version version() { return HttpClient.Version.HTTP_1_1; }
    }
}
